[BasicAA] Handle scalable type sizes with constant offsets

davemgreen opened the pull request:
This is a separate, but related issue to #69152 that was attempting to improve
AA with scalable dependency distances. This patch attempts to improve when
there are scalable accesses with a constant offset between them. We happen to
get a report of such a thing recently, where so long as the vscale_range is
known, the maximum size of the access can be assessed and better aliasing
results can be returned.

The Upper range of the vscale_range, along with known part of the typesize are
used to prove that Off >= CR.upper * LSize. It does not try to produce
PartialAlias results at the moment from the lower vscale_range. It also enables
the added benefit of allowing better alias analysis when the RHS of the two
values is scalable, but the LHS is normal and can be treated like any other
aliasing query.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1146,24 +1142,37 @@ AliasResult BasicAAResult::aliasGEP(
-      return AR;
+      return AliasResult::NoAlias;
+    } else {
+      // We can used the getVScaleRange to prove that Off >= (CR.upper * LSize).
```
```suggestion
      // We can use the getVScaleRange to prove that Off >= (CR.upper * LSize).
```

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1146,24 +1142,37 @@ AliasResult BasicAAResult::aliasGEP(
+      // We can used the getVScaleRange to prove that Off >= (CR.upper * LSize).
+      ConstantRange CR = getVScaleRange(&F, Off.getBitWidth());
+      APInt UpperRange = CR.getUnsignedMax().umul_sat(
+          APInt(Off.getBitWidth(), LSize.getKnownMinValue()));
```
It may be cleaner to express this using umul_ov and then `!Overflow` rather than umul_sat.

nikic reviewed (approved):
LGTM

davemgreen committed:
[BasicAA] Handle scalable type sizes with constant offsets

This is a separate, but related issue to #69152 that was attempting to improve
AA with scalable dependency distances. This patch attempts to improve when
there are scalable accesses with a constant offset between them. We happen to
get a report of such a thing recently, where so long as the vscale_range is
known, the maximum size of the access can be assessed and better aliasing
results can be returned.

The Upper range of the vscale_range, along with known part of the typesize are
used to prove that Off >= CR.upper * LSize. It does not try to produce
PartialAlias results at the moment from the lower vscale_range.
